### PR TITLE
Natively support Plotly figure in the chart control

### LINKED ISSUE #640: Is it possible to show plotly.graph_objscts Figure?
I have code like below.
Is it possible to show plotly.graph_objscts Figure?

```
fig1 = go.Figure(
            data=go.Contour(
                z=O,
                x=p,
                y=y,
                colorscale="Jet",
                contours=dict(start=1.0, end=6.0, size=0.5),
                colorbar=dict(tick0=1.5, dtick=0.5, x=1.08, len=0.7),
            )
        ) 
fig1.add_trace(
        go.Scatter(
                x=[p],
                y=[C],
                mode="markers",
                marker=dict(color="white", symbol="circle", size=10, line=dict(color="black", width=1)),
                name="Cur",
            )
        )
fig1.update_layout(xaxis_title="P", yaxis_title="C")

```

### LINKED ISSUE #653: Other integration for Plotly Python in Taipy
**Description**
The goal of this issue is to improve the integration of Plotly Python into Taipy to explore other ways to do it.

This issue is related to: https://github.com/Avaiga/taipy/issues/640, https://github.com/Avaiga/taipy/issues/570

A way could be for example, to use the Plotly object to extract the different parameters to generate the chart. A Python Plotly figure can be translated into *json* and much more.

Here is a class that tries to provide a mapping from the Python Plotly figure to Taipy.
*tp_plotly.py*:
```python
import json

class PlotlyTaipyChart:
    def __init__(self, fig):
        self.layout = json.loads(fig.layout.to_json())
        self.traces = {}
        self.title = fig.layout.title.text
        self.data = []
        
        AXES_TYPES = ['x', 'y', 'z', 'lon', 'lat', 'locations',
                      'theta', 'r', 'values', 'labels',
                      'parents', 'base', 'measure', 'close', 'open', 'low', 'high', 'text']
        
        for i, trace in enumerate(fig.data):
            data_trace = {}
            dict_trace = json.loads(trace.to_json())
            new_trace = {f'{key}[{i+1}]':value for key, value in dict_trace.items() if key not in AXES_TYPES+['marker', 'hovertemplate']}
            self.traces.update(new_trace)

            # Add traces and columns into the data property of the Taipy Chart
            for key, value in dict_trace.items():
                if key in AXES_TYPES:
                    data_trace[f'{key}[{i+1}]'] = value
                    self.traces[f'{key}[{i+1}]'] = f'{i}/{key}[{i+1}]'
                elif key == 'marker':
                    self.traces[f'marker[{i+1}]'] = value
                    if 'color' in dict_trace['marker']:
                        if not isinstance(value['color'], str):
                            data_trace[f'color[{i+1}]'] = value['color']
                            self.traces[f'marker[{i+1}]']["color"] = f'{i}/color[{i+1}]'
                    if 'size' in dict_trace['marker']:
                        if not isinstance(value['size'], str):
                            data_trace[f'size[{i+1}]'] = value['size']
                            self.traces[f'marker[{i+1}]']["size"] = f'{i}/size[{i+1}]'
                            
            self.data.append(data_trace)
            self.traces[f'options[{i+1}]'] = dict_trace

    def to_taipy(self):
        """
        Returns a dictionary that can be used by Taipy
        """
        taipy_dict = {
            'layout': self.layout,
            'data': self.data,
            'title': self.title
            }
        taipy_dict.update(self.traces)
        return taipy_dict
    
    
    def to_taipy_json(self, file_name):
        with open(file_name, 'w') as f:
            json.dump(self.to_taipy(), f, indent=4)


```

Here is a script that showcases that this mapping works for a lot of common charts:
```python
from tp_plotly import PlotlyTaipyChart

import plotly.graph_objects as go

import pandas as pd
from datetime import datetime

# -------------------------------CANDLE------------------------------

df = pd.read_csv('https://raw.githubusercontent.com/plotly/datasets/master/finance-charts-apple.csv')

candle_fig = go.Figure(data=[go.Candlestick(x=df['Date'],
                open=df['AAPL.Open'],
                high=df['AAPL.High'],
                low=df['AAPL.Low'],
                close=df['AAPL.Close'])])

# ----------------------------BAR and SCATTER---------------------------------

# Create random data with numpy
import numpy as np
np.random.seed(1)

N = 100
random_x = np.linspace(0, 1, N)
random_y0 = np.random.randn(N) + 5
random_y1 = np.random.randn(N)
random_y2 = np.random.randn(N) - 5

# Create traces
bar_line_fig = go.Figure()
bar_line_fig.add_trace(go.Scatter(x=random_x, y=random_y0,
                    mode='lines',
                    name='lines'))
bar_line_fig.add_trace(go.Scatter(x=random_x, y=random_y1,
                    mode='lines+markers',
                    name='lines+markers'))
bar_line_fig.add_trace(go.Scatter(x=random_x, y=random_y2,
                    mode='markers', name='markers'))
bar_line_fig.add_trace(go.Bar(x=random_x, y=random_y2))

# ------------------------------PIE-------------------------------

import plotly.express as px
df = px.data.gapminder().query("year == 2007").query("continent == 'Europe'")
df.loc[df['pop'] < 2.e6, 'country'] = 'Other countries' # Represent only large countries
pie_fig = px.pie(df, values='pop', names='country', title='Population of European continent')


# ---------------------------BUBBLE-----------------------------
df = px.data.gapminder()

bubble_fig = px.scatter(df.query("year==2007"), x="gdpPercap", y="lifeExp",
	         size="pop", color="continent",
                 hover_name="country", log_x=True, size_max=60)


# ---------------------------HISTOGRAM-----------------------------


import plotly.express as px
df = px.data.tips()
histogram_fig = px.histogram(df, x="total_bill", color="sex")


# ---------------------------MAP-----------------------------

df = px.data.election()
geojson = px.data.election_geojson()

choropleth_fig = px.choropleth_mapbox(df, geojson=geojson, color="winner",
                           locations="district", featureidkey="properties.district",
                           center={"lat": 45.5517, "lon": -73.7073},
                           mapbox_style="carto-positron", zoom=9)
choropleth_fig.update_layout(margin={"r":0,"t":0,"l":0,"b":0})

# ---------------------------CONVERSION-----------------------------

bar_line_taipy = PlotlyTaipyChart(bar_line_fig)
pie_taipy = PlotlyTaipyChart(pie_fig)
candle_taipy = PlotlyTaipyChart(candle_fig)
bubble_taipy = PlotlyTaipyChart(bubble_fig)
histogram_taipy = PlotlyTaipyChart(histogram_fig)
choropleth_taipy = PlotlyTaipyChart(choropleth_fig)

# --------------------------------------------------------------------

from taipy.gui import Gui 

md = """
<|chart|properties={bar_line_taipy.to_taipy()}|>

<|chart|properties={pie_taipy.to_taipy()}|>

<|chart|properties={candle_taipy.to_taipy()}|>

<|chart|properties={bubble_taipy.to_taipy()}|>

<|chart|properties={histogram_taipy.to_taipy()}|>

<|chart|properties={choropleth_taipy.to_taipy()}|>
"""

Gui(md).run()
```

image

image

image

This method potentially provides full integration of Python Plotly in Taipy with fewer performance issues, the same layout, and the ability to get events from the chart (selection).

**Implementation proposal**

One way to implement this in a very light manner (from the user stand point) would be to add a new property to the chart control that accepts a `plotly.graph_objects._figure.Figure` instance. This instance could then provide its JSON configuration, that would be impacted by the potential value of the `properties` property, then each individual chart property, clarifying what the property priorities are.
The resulting configuration could then be used by the Chart component for rendering (and interation).

**Acceptance Criteria**
- [ ] Ensure new code is unit tested, and check code coverage is at least 90%
- [ ] Propagate any change on the demos and run all of them to ensure there is no breaking change
- [ ] Ensure any change is well documented. This is handled in [this issue](https://github.com/Avaiga/taipy-doc/issues/779).

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
Resolves #653.
Resolves #640.

```
candle_fig = go.Figure(
    data=[
        go.Candlestick(
            x=df["Date"], open=df["AAPL.Open"], high=df["AAPL.High"], low=df["AAPL.Low"], close=df["AAPL.Close"]
        )
    ]
)

<|chart|figure={candle_fig}|>
```
image

### CODE REVIEW

**Review by dinhlongviolin1 — DISMISSED** [review]
Much simpler than expected!

**Review by dinhlongviolin1 — DISMISSED** [review]

File: taipy/gui/utils/types.py [hunk 1]
```diff
@@ -186,3 +188,25 @@ def get(self):
...
+        if not val:
+            return None
+        if find_spec("plotly") and find_spec("plotly.graph_objs"):
+            from plotly.graph_objs import Figure as PlotlyFigure
+
+            if isinstance(val, PlotlyFigure):
+                try:
+                    return [json.loads(val.to_json())]
+                except Exception as e:
+                    _warn("Issue while serailizing Plotly Figure", e)
```
Comment (FabienLelaquais) [inline taipy/gui/utils/types.py]:
```suggestion
                    _warn("Issue while serializing Plotly Figure", e)
```

File: taipy/gui/utils/types.py [hunk 2]
```diff
@@ -186,3 +188,25 @@ def get(self):
...
+                    return None
+        _warn("chart.figure should be of type Plotly Figure")
```
Comment (FabienLelaquais) [inline taipy/gui/utils/types.py]:
```suggestion
        _warn("'figure' property value should be a plotly.graph_objects.Figure.")
```
or maybe 'must'

File: taipy/gui/viselements.json [hunk 3]
```diff
@@ -554,6 +554,11 @@
             "type": "dynamic(bool)",
             "default_value": "False",
             "doc": "Allows dynamic config refresh if set to True."
+          },
+          {
+            "name": "figure",
+            "type": "Plotly Figure",
```
Comment (FabienLelaquais) [inline taipy/gui/viselements.json]:
```suggestion
            "type": "plotly.graph_objects.Figure",
```

**Review by FabienLelaquais — DISMISSED** [review]
👌👌👌

**Review by FabienLelaquais — APPROVED** [review]
👌

**Review by dinhlongviolin1 — APPROVED** [review]

**Pull request merged**